{"cells":[{"cell_type":"markdown","metadata":{"id":"3IkqRycJWOCn"},"source":["**Import needed packages/modules**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Jd4lLFM-WMua"},"outputs":[],"source":["# Cell 1\n","import re\n","from pathlib import Path\n","\n","from google.colab import drive"]},{"cell_type":"markdown","metadata":{"id":"kjX9F_QhXlg4"},"source":["**Mount your Google Drive and determine the path to this notebook**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"4ckdruqrXvTp"},"outputs":[],"source":["# Cell 2\n","drive.mount(\"/content/gdrive\", force_remount=True)\n","notebook_path = Path(\"/content/gdrive/MyDrive/SciComp101-GC\")\n","notebook_path /= Path(\"Session 15 - Coding for Biology and Chemistry\")\n","notebook_path"]},{"cell_type":"markdown","metadata":{"id":"suse3Q1wz_t5"},"source":["**Create a dictionary `AMINO_ACIDS`**\n","1. Each key has two string values: The 3-letter amino acid <u>code</u>, and the 1-letter amino acid <u>symbol</u>\n","2. Each value is a Python `list` containing the various <u>codons</u> the encode that associated amino acid"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"JIE6SoCc1BlX"},"outputs":[],"source":["# Cell 3\n","AMINO_ACIDS = {\n","    (\"Ala\", \"A\"): [\"GCT\", \"GCA\", \"GCC\", \"GCG\"],  # Alanine\n","    (\"Arg\", \"R\"): [\"CGT\", \"CGC\", \"CGA\", \"CGG\", \"AGA\", \"AGG\"],  # Arginine\n","    (\"Asn\", \"N\"): [\"AAT\", \"AAC\"],  # Asparagine\n","    (\"Asp\", \"D\"): [\"GAT\", \"GAC\"],  # Aspartic Acid\n","    (\"Cys\", \"C\"): [\"TGT\", \"TGC\"],  # Cysteine\n","    (\"Gln\", \"Q\"): [\"CAA\", \"CAG\"],  # Glutamine\n","    (\"Glu\", \"E\"): [\"GAA\", \"GAG\"],  # Glutamic Acid\n","    (\"Gly\", \"G\"): [\"GGT\", \"GGC\", \"GGA\", \"GGG\"],  # Glycine\n","    (\"His\", \"H\"): [\"CAT\", \"CAC\"],  # Histidine\n","    (\"Ile\", \"I\"): [\"ATT\", \"ATC\", \"ATA\"],  # Isoleucine\n","    (\"Leu\", \"L\"): [\"TTA\", \"TTG\", \"CTT\", \"CTC\", \"CTA\", \"CTG\"],  # Leucine\n","    (\"Lys\", \"K\"): [\"AAA\", \"AAG\"],  # Lysine\n","    (\"Met\", \"M\"): [\"ATG\"],  # Methionine (Start)\n","    (\"Phe\", \"F\"): [\"TTT\", \"TTC\"],  # Phenylalanine\n","    (\"Pro\", \"P\"): [\"CCT\", \"CCC\", \"CCA\", \"CCG\"],  # Proline\n","    (\"Ser\", \"S\"): [\"TCT\", \"TCC\", \"TCA\", \"TCG\", \"AGT\", \"AGC\"],  # Serine\n","    (\"Thr\", \"T\"): [\"ACT\", \"ACC\", \"ACA\", \"ACG\"],  # Threonine\n","    (\"Trp\", \"W\"): [\"TGG\"],  # Tryptophan\n","    (\"Tyr\", \"Y\"): [\"TAT\", \"TAC\"],  # Tyrosine\n","    (\"Val\", \"V\"): [\"GTT\", \"GTC\", \"GTA\", \"GTG\"],  # Valine\n","    (\"Stop\", \"0\"): [\"TAA\", \"TGA\", \"TAG\"],  # Stop\n","}\n","display(AMINO_ACIDS)"]},{"cell_type":"markdown","metadata":{"id":"E0H0E2fB16et"},"source":["**Define a function to build and return the <u>reverse complement</u> of a DNA sequence**\n","1. Use a `for` loop to retrieve - in <u>reverse</u> order - each character in the given sequence\n","2. Build the reverse complement string by matching **A<->T** and **C<->G**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"AHD9gIf321ls"},"outputs":[],"source":["# Cell 4\n","def reverse_complement(seq):\n","    seq_reverse = \"\"\n","    for c in seq[::-1]:\n","        if c == \"C\":\n","            seq_reverse += \"G\"\n","        elif c == \"G\":\n","            seq_reverse += \"C\"\n","        elif c == \"A\":\n","            seq_reverse += \"T\"\n","        elif c == \"T\":\n","            seq_reverse += \"A\"\n","    return seq_reverse\n","\n","reverse_complement(\"ATCG\")"]},{"cell_type":"markdown","metadata":{"id":"Xsr7H8PJ3gl8"},"source":["**Define a function to return a Python `list` of <u>codons</u> based upon a given sequence and starting *offset* in that sequence**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"eBL0kK613z5t"},"outputs":[],"source":["# Cell 5\n","def make_codons(seq, offset):\n","    codons = []\n","    # Each codon is a grouping of three successive nucleotides,\n","    # starting at some offset in a DNA/RNA sequence\n","    # Each codon encodes a specific amino acid\n","    for i in range(offset, len(seq), 3):\n","        if i + 3 <= len(seq):\n","            codons.append(seq[i : i + 3])\n","    return codons\n","\n","\n","print(make_codons(\"ATCGGAT\", 0))\n","print(make_codons(\"ATCGGAT\", 1))\n","print(make_codons(\"ATCGGAT\", 2))"]},{"cell_type":"markdown","metadata":{"id":"u1hgkuS34h16"},"source":["**Define a function to return the index of a given codon in a `list` of codons**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"7t6xn41-4vEN"},"outputs":[],"source":["# Cell 6\n","def find_codon(codon_list, codon):\n","    try:\n","        # Get index if this codon appears in the codon list\n","        idx = codon_list.index(codon)\n","        return idx\n","    except ValueError:\n","        # This codon does not appear in the list\n","        return -1\n","\n","\n","print(find_codon(make_codons(\"ATCGGAT\", 0), \"GGA\"))\n","print(find_codon(make_codons(\"ATCGGAT\", 1), \"TCG\"))\n","print(find_codon(make_codons(\"ATCGGAT\", 2), \"GAT\"))"]},{"cell_type":"markdown","metadata":{"id":"ZZ__n0QD6JmU"},"source":["**Define a function that returns the string of 1-letter amino acid symbols for a given codon string**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"JzUuWHUy6cP8"},"outputs":[],"source":["# Cell 7\n","def decode_codons(codon_string):\n","    # Invert AMINO_ACIDS so it is keyed by codon, not amino acid\n","    inverted_dict = {}\n","    for k in AMINO_ACIDS:\n","        for v in AMINO_ACIDS[k]:\n","            inverted_dict[v] = k\n","    # Build string of single-letter amino acids based upon each codon\n","    acids = \"\"\n","    for c in codon_string.split():\n","        # The single letter is the 2nd element (aka [1]) in each key's tuple\n","        acids += inverted_dict[c][1]\n","    # Don't include start and stop codons in amino acid sequence\n","    acids = acids[1:-1]\n","    return acids\n","\n","\n","decode_codons(\"ATG GGA GTC GAT TAG\")"]},{"cell_type":"markdown","metadata":{"id":"QvYyLDCo7KnW"},"source":["**Define a function to get the codons within a valid <u>open reading frame<u>**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"_8VXSkms7dVU"},"outputs":[],"source":["# Cell 8\n","def get_orf(codon_string, offset):\n","    # Split sequence into codon list starting at offset\n","    codon_list = make_codons(codon_string, offset)\n","\n","    # Find possible index for START codon\n","    start_idx = find_codon(codon_list, \"ATG\")\n","    if start_idx < 0:\n","        return None\n","\n","    # Find possible indexes for all three STOP codons\n","    stop_indexes = [\n","        find_codon(codon_list, \"TAA\"),\n","        find_codon(codon_list, \"TAG\"),\n","        find_codon(codon_list, \"TGA\"),\n","    ]\n","\n","    # Remove any STOP codon index that comes before the START index\n","    stop_indexes[:] = [idx for idx in stop_indexes if idx > start_idx]\n","    if len(stop_indexes) == 0:\n","        return None\n","\n","    # Use the index of the first occurring valid STOP codon\n","    stop_idx = min(stop_indexes)\n","\n","    # An empty frame if STOP immediately follows START\n","    if stop_idx == start_idx + 1:\n","        return None\n","\n","    # Build a string of all codons (including START and STOP)\n","    codon_string = \"\"\n","    for idx in range(start_idx, stop_idx + 1):\n","        codon_string += codon_list[idx] + \" \"\n","\n","    # If an ORF exists, also display codons as single character amino acids\n","    if len(codon_string) > 0:\n","        codon_string += f\" ({decode_codons(codon_string)})\"\n","\n","    return codon_string\n","\n","print(get_orf(\"ATGGGAGTCGATTAG\", 0))\n","print(get_orf(\"ATGGGAGTCGATTAG\", 1))"]},{"cell_type":"markdown","metadata":{"id":"Z5JPKFU-8N62"},"source":["**Define a function to find all the possible ORFs within a given ASCII text sequence file**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"W_DPntCU8blM"},"outputs":[],"source":["# Cell 9\n","def analyze_file(file_name):\n","    file_path = notebook_path / file_name\n","    with Path.open(file_path, \"rb\") as f_in:\n","        # Read in text file into an array of file bytes\n","        f_bytes = bytearray(f_in.read())\n","\n","    # Enforce uppercase and remove non-letters, convert to UTF-8\n","    seq = bytearray(f_bytes).decode().upper()\n","    seq = re.compile(\"[^A-Z]\").sub(\"\", seq)\n","\n","    # Print the original given sequence\n","    print(f\"Original sequence:  {seq} \")\n","\n","    # Print any open reading frames in forward sequence\n","    if s := get_orf(seq, 0):\n","        print(f\"Original sequence Open Frame +0 : {s}\")\n","    if s := get_orf(seq, 1):\n","        print(f\"Original sequence Open Frame +1 : {s}\")\n","    if s := get_orf(seq, 2):\n","        print(f\"Original sequence Open Frame +2 : {s}\")\n","\n","    # Build and print the reverse compliment of the given sequence\n","    seq_rc = reverse_complement(seq)\n","    print(f\"Reverse complement: {seq_rc} \")\n","\n","    # Print any open reading frames in reverse compliment of sequence\n","    if s := get_orf(seq_rc, 0):\n","        print(f\"Reverse complement Open Frame +0: {s}\")\n","    if s := get_orf(seq_rc, 1):\n","        print(f\"Reverse complement Open Frame +1: {s}\")\n","    if s := get_orf(seq_rc, 2):\n","        print(f\"Reverse complement Open Frame +2: {s}\")"]},{"cell_type":"markdown","metadata":{"id":"-qBM8p_K849F"},"source":["**Analyze the sequence text file \"seq2.txt\"**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ZEK7e6OE8-YJ"},"outputs":[],"source":["# Cell 10\n","analyze_file(\"seq2.txt\")"]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python","version":"3.12.1"}},"nbformat":4,"nbformat_minor":0}